### ISSUE TITLE
1.9.0 - `no-unused-vars` diagnosis is wrong in cases where the value is accessible via a closure

### ISSUE THREAD

**Issue by aricart** [denoland/deno_lint#667]
1.9.0 complaints about values that will be found in the closure, here's an example:

```typescript
const i = setInterval(async () => {
  try {
    const v = await fn();
    if (v) {
      clearInterval(i);
      done(v);
    }
  } catch (_) {
    // ignore
  }
}, opts?.interval);
```

The linter will:

> (no-unused-vars) `i` is never used
>    const i = setInterval(async () => {

**Comment by magurotuna** [comment]
@ebebbington Thanks for additional info. That looks a little bit different from what @aricart showed, but is indeed a false positive. I'll try to fix it too.

Looking at the rule implementation, both of those cases will require much rework to fix. It'll likely take a while.

**Comment by Soremwar** [comment]
An easier reproduction

```ts
function variable(){
  // Variable is unused and never redefined according to the linter
  let variable = 1;

  variable = 2;
}

variable();
```

**Comment by KyleJune** [comment]
I use a class as a mixin, copying all of its methods onto objects. Before I specified the type for this in the toString function's signature. Now it shows this as unused.

```ts
  interface Point4D extends Point, TimePoint {
    (x: number, y: number, z: number, time: number): string;
    z: number;
    toString(): string;
  }
  abstract class Point4DPartial {
    toString(this: Point4D): string {
      return [this.getPosition(), this.z, this.getTime()].join(", ");
    }
  }
  ```

**Comment by magurotuna** [comment]
@Soremwar Seems like the linter emits the right error in your example. `variable` is declared and then mutated to `2`, but _never read out_. Variables that will never be read out are thought of as "unused".

@KyleJune Thanks for the info, that's a bug of the linter. I'll look into it.

**Comment by Soremwar** [comment]
@magurotuna I'm sorry, wrong reproduction

```ts
setTimeout(function cidr() {
  // I have to rename either the function or the variable to make the error go away
  // This will happen with any function that receives a named callback as an argument
  const cidr = "192.168.100.128/25"; 
  
  console.log(cidr)
});
```

**Comment by magurotuna** [comment]
@bartlomieju oh sorry, #678 doesn't fix all issues reported here. I'll tackle the rest ones.
